Refactor geoip cache for MaxMind databases

joegallo opened the pull request:
This PR adds another degree of freedom in the `MaxmindIpDataLookups`. Rather than requiring that the response from the `Reader` is necessarily the same object as we keep in the `GeoIpCache`, it adds a new `record(...)` method (read as a noun, like "the record for storing in the cache") that allows a given MaxMind `IpDataLookup` to build a different object out of the response and then cache **that** object.

In this PR, only the `CountryResponse` is refactored to use this new capability -- it's the smallest of the response objects that would benefit from a change, so it's the easiest one to review at the same time as the rest of the machinery changes.

We'll benefit from this new capability in two ways: 
- first, it will make cache hits faster -- we can update an `IngestDocument` from a record faster than we can update from a proper `Response` (as a simple example, the common `getName()` method of various responses iterates over a collection -- now we'll do that just once when we populate the cache rather than every time we have a cache hit)
- second, it decreases the size of the objects in the cache by about a factor of ten, so in some subsequent PR, we could reasonably expect to cache way more of these objects (and have more cache hits as a result)

Or, to put it simply, things are faster and use less memory this way.

joegallo committed:
Introduce another layer of indirection

Allow a given lookup to create a cache-able record from the response
from the database. At the moment every lookup just uses its response
from the database as the cache-able record (this is fine).

joegallo committed:
Use a custom cache record for CountryResponse

joegallo committed:
Silence a warning from IntelliJ

joegallo committed:
This variable cannot be null

CityResponse and EnterpriseResponse use the null-object pattern here,
so postal is never null.

joegallo committed:
Promote the Result helper record to IpDataLookup

so that it can be reused. Of course that means that we have to tweak
all the existing callers to use method access rather than field
access, hence the bigger diff than one might otherwise imagine.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

masseyke commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/MaxmindIpDataLookups.java:
```diff
@@ -116,6 +118,11 @@ static class AnonymousIp extends AbstractBase<AnonymousIpResponse> {
         }
 
+        @Override
+        protected AnonymousIpResponse record(AnonymousIpResponse response) {
```
I kind of wonder if this method name is going to cause confusion for future maintainers, since as you've pointed out your first thought when you see this is the verb -- what am I recording? What about `cacheableRecord`?

joegallo replied:
Indeed, my unwavering love of short names is definitely at odds with the desire for a clear name in this case. I renamed it to `cacheableRecord` for now via 9dd8964, but I think I'm going to polish things a little more (on subsequent PRs).

masseyke reviewed (approved):
LGTM

joegallo committed:
Twiddle some docstrings

joegallo committed:
Lampshade a small oddity

joegallo committed:
Rename these variables

There's too many nearby things named record or result or response. So
here's a new word, 'entry'. A database has a bunch of entries in it.

joegallo committed:
Merge branch 'main' into refactor-ingest-geoip-maxmind-cache

joegallo committed:
Rename this method

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/125693)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/125693">](https://github.com/elastic/elasticsearch/pull/125693) |
